scapy/contrib/automotive/can_xcp.py

```diff
@@ -0,0 +1,302 @@
+        self.__cycle_time = cycle_time
+
+    def run(self):
+        """
+        Main funciton in send thread
+        Sends the messeage set in Constructor every 0.02 Seconds
+        and increases the id of the msg after each send
+        until __end is reached
+        """
+        self.__console.println("Start Sending")
+        for i in range(self.__start, self.__end + 1):
+            self.__message.identifier = i
+            self.__current_id_var.c.acquire()
+            self.__current_id_var.id = i
+            self.__current_id_var.c.release()
+            self.__socket.send(self.__message)
+            time.sleep(self.__cycle_time)
+        self.__console.println("Done")
+
+
+class XCP_SNIFF_THREAD(threading.Thread):
+    """
+    Reads data from the can bus and checks if a XCP Response
+    has been sent on the CANBus
+    """
+    def __init__(self, can_socket,
+                 current_id_var, result_list, known_ids, console):
+        """
+        Constructor
+        :param can_socket: Socket used for sending messages
+        :param current_id_var: Shard variable that holds the last
+            used ID set by the sending thread
+        :param result_list: list of detected XCP-Connections
+        :param konwn_ids: Noise message IDs sent during sniff.
+            Holds all message IDs sent during the first
+            3 seconds at Bus startup
+        """
+        threading.Thread.__init__(self)
+        self.__socket = can_socket
+        if isinstance(known_ids, list):
+            self.__known_ids = known_ids
+        else:
+            self.__known_ids = list()
+        self.__current_id_var = current_id_var
+        self.__result_list = result_list
+        self._stop_event = threading.Event()
+        self.__console = console
+
+    def set_known_ids(self, ids):
+        self.__known_ids = ids
+
+    def run(self):
+        """
+        Main Funciton of sniffer:
+        Receives all message sent on CAN. If a XCP response has been
+            received a result is added
+        to the result list
+        """
+        data_array = list()
+        self.__console.println("Start Sniffing")
+        while not self._stop_event.is_set():
+            try:
+                msg = self.__socket.recv()
+            except TypeError:
+                continue
+            if msg:
+                data_array = [x for x in msg.data]
+            if (msg and data_array and data_array[0] == 0xFF and
+                    msg.identifier not in self.__known_ids and
+                    msg.length == 8 and
+                    msg.identifier != self.__current_id_var.id):
+                self.__console.println(hex(msg.identifier) +
+                                       " " +
+                                       hex(self.__current_id_var.id))
+                request_id = self.__current_id_var.id
+                response_id = msg.identifier
+                max_dto = (data_array[4] << 8) + data_array[5]
```

Comment:
The interpretation of data should be done through `Packet` definitions. For example: 
```python
class myPacket(Packet):
    field_des = [
        ... some fields ...
    ]

some_socket.basecls = myPacket
# or
raw_data = some_socket.recv() # or some_socket.sniff(...)
pkt = myPacket(raw_data)
pkt.someField == whatever
```